### DISCUSSION TITLE
bug: Procedure return types result in `any` when using non-relative import paths

### DISCUSSION THREAD

**Discussion by delyan-peev** [trpc/trpc#3476] [Q&A]
### Provide environment information

```
  System:
    OS: macOS 12.5
    CPU: (10) arm64 Apple M1 Pro
    Memory: 2.67 GB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh

  Binaries:
    Node: 16.18.1 - ~/.nvm/versions/node/v16.18.1/bin/node
    npm: 8.19.2 - ~/.nvm/versions/node/v16.18.1/bin/npm

  Browsers:
    Chrome: 108.0.5359.124
    Safari: 15.6

  npmPackages (Server):
    @trpc/server: ^10.4.3 => 10.4.3 
    typescript: ^4.9.4 => 4.9.4

  npmPackages (Client):
    @tanstack/react-query: ^4.19.1 => 4.19.1 
    @trpc/client: ^10.4.3 => 10.4.3 
    @trpc/react-query: ^10.4.3 => 10.4.3 
    @trpc/server: ^10.4.3 => 10.4.3 
    react: ^18.2.0 => 18.2.0 
    typescript: ^4.9.4 => 4.9.4
```

### Describe the bug

Hey guys,

First of all, thanks for the awesome package, it's been revolutionizing the way we write React apps. 

Not sure if this is a known issue, but return types of queries (and not only) in the client result in `any` when using non-relative (either absolute or TS aliased paths) imports on the server. I supposed it is a bug, since I did not find the docs mentioning relative paths are required.

Actually, if you change the import of the whole `entitiesRouter` to be non-relative (see example and description below), it will return an object of type `any` for `trpc.entities` in the client. So the issue is not only for queries but for anything that is not imported with a relative path in `trpc`.

This is kind of frustrating for us, since it requires us to use relative imports pretty much everywhere in the project and it can easily result in a relative imports hell.

E.g.
`import { A } from '../../../../../../wherever/whatever';`

Is there a workaround for this and what would it be?

Apologies if this has been asked and answered, I searched through the issues and the docs, but did not find anything.

Regards,
Delyan

### Link to reproduction

https://github.com/delyan-peev/trpc-relative-paths-bug

### To reproduce

1. Run `npm install` in both the `api-server` and `client` directories.
2. Run `npx prisma generate client` in the `api-server` directory.
3. Observe the returned type of the data property in `client/src/Test.tsx`
4. Change the `import { prisma } from 'prisma';` line in `api-server/routers/entitiesRouter.ts` to `import { prisma } from '../prisma';`
5. Observe the returned type of the data property in `client/src/Test.tsx`

### 👨‍👧‍👦 Contributing

- [x] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment]
Thanks for the kind words!

This is a tsconfig issue or that you have different versions across packages or similar, I don't have time to look into it 🤶 but maybe someone else can provide guidance

**Reply by delyan-peev** [reply]
I suspect it is something related to the server's tsconfig, however I'm still failing to find exactly what it is. Anyway, I appreciate your answer on this very day. I'd also be extremely grateful if someone can take a look once all the holiday buzz has passed.

Happy holidays everyone!

**Comment by delyan-peev** [comment]
Update: It seems to be something with `prisma` as if I change the return type to not be something of prisma's entities (e.g. a hardcoded `{name: 'somename'}` object), it seems to work just fine.

**Reply by KATT** [reply]
Google prisma nohoist

**Reply by delyan-peev** [reply]
I've just tried with both hoisted and non-hoisted prisma (and all other packages) and the problem still persists. I doubt this is the issue as the example in the repo has no "root" package per se, so prisma cannot be hoisted to any upper-directory node_modules.

I modified it to have a root package and declare the respective workspaces to test the hoisting theory. Unfortunately, to no avail :(

**Reply by delyan-peev** [reply]
P.S. I probably understood what you meant a minute too late. Issue is fixed now leveraging [TypeScript's project references](https://www.typescriptlang.org/docs/handbook/project-references.html) feature.

Read more about the fix in [this](https://twitter.com/colinhacks/status/1298030688063561728?lang=en) slightly ranty tweet by Colin.

Thanks for the help and have a great new year!
Hope this helps someone in the future. I can finally sleep now 💤 😃

**Reply by sunnhyn** [reply]
@delyan-peev Thank you! You've saved me a lot of time! 👍🏻 Now I can finally sleep too. 😃
